File: SymbolCollector.py
from AdvancedBaseVisitor import AdvancedBaseVisitor
from typing import Optional, List, Tuple, Dict
from Symbol import (
    VariableSymbol,
    VariableType,
    DatatypeSymbol,
    GenericPlaceholderSymbol,
)
from FunctionSymbol import FunctionSymbol
from Error import CompilerError, InternalError, ImpossibleSituation
from Scope import Scope
from grammar.HazeParser import HazeParser
from Datatype import Datatype, FunctionLinkage
from Program import Program, ExternFunctionRef
from utils import resolveGenerics


class SymbolCollector(AdvancedBaseVisitor):
    def __init__(self, filename, db):
        super().__init__(filename, db)
        self.program = Program(db)
        self.currentLinkage: Optional[FunctionLinkage] = None
        self.structStack: List[DatatypeSymbol] = []
        self.currentFunctionSymbols: List[FunctionSymbol] = []

    def visitParam(self, ctx):
        return ctx.ID().getText(), self.visit(ctx.datatype())

    def visitGenericDatatype(self, ctx: HazeParser.GenericDatatypeContext):
        name = ctx.ID().getText()
        foundSymbol = self.db.getCurrentScope().tryLookupSymbol(
            name, self.getLocation(ctx)
        )
        if not foundSymbol:
            raise CompilerError(
                f"Type '{name}' is not defined.",
                self.getLocation(ctx),
            )

        genericsProvided = [self.visit(n) for n in ctx.datatype()]
        if len(foundSymbol.type.generics) != len(genericsProvided):
            raise CompilerError(
                f"Datatype expected {len(foundSymbol.type.generics)} generic arguments but got {len(genericsProvided)}.",
                self.getLocation(ctx),
            )
        dt = resolveGenerics(
            foundSymbol.type, self.db.getCurrentScope(), self.getLocation(ctx)
        )
        return dt

    def visitFunctionDatatype(self, ctx):
        params = []
        for param in ctx.functype().params().param():
            params.append(self.visit(param))

        return Datatype.createFunctionType(
            params,
            self.visit(ctx.functype().returntype()),
        )

    def visitExternblock(self, ctx):
        lang = ctx.externlang().getText()[1:-1]
        match lang:
            case "C":
                if self.currentLinkage:
                    raise CompilerError(
                        "Extern blocks cannot be nested",
                        self.getLocation(ctx),
                    )
                self.currentLinkage = FunctionLinkage.External_C
                self.visitChildren(ctx)
                self.currentLinkage = None
            case _:
                raise CompilerError(
                    f"Extern Language '{lang}' is not supported.",
                    self.getLocation(ctx.externlang()),
                )

    # def visitCompilationhint(self, ctx):
    #     self.visitChildren(ctx)
    #     lang = ctx.compilationlang().getText()[1:-1]
    #     if lang != "C":
    #         raise CompilerError(
    #             f"Compilation Language '{lang}' is not supported.",
    #             self.getLocation(ctx),
    #         )

    #     filename = ctx.compilationhintfilename().getText()[1:-1]
    #     flags: str = ""
    #     if ctx.compilationhintflags():
    #         flags = ctx.compilationhintflags().getText()[1:-1]

    #     self.db.defineExternalCompilationUnit(filename, lang, flags.split(" "))
    #     self.setNodeCompilationHintFilename(ctx, filename)

    # def visitLinkerhint(self, ctx):
    #     self.useCurrentNodeScope(ctx)
    #     self.visitChildren(ctx)
    #     self.db.addExternalLinkerFlags(ctx.STRING_LITERAL().getText().split(" "))

    def visitReturntype(self, ctx):
        return self.visit(ctx.datatype())

    def visitExternfuncdef(self, ctx: HazeParser.ExternfuncdefContext):
        if not self.currentLinkage:
            raise ImpossibleSituation()

        signature = [n.getText() for n in ctx.ID()]
        if len(signature) > 1 and self.currentLinkage == FunctionLinkage.External_C:
            raise CompilerError(
                "Extern C functions cannot be namespaced", self.getLocation(ctx)
            )

        if len(signature) > 1:
            raise InternalError(
                "Namespacing for external function is not implemented yet"
            )

        name = signature[-1]
        symbol = FunctionSymbol(
            name,
            None,
            Datatype.createFunctionType(
                [], Datatype.createPrimitiveType(Datatype.PrimitiveVariants.none)
            ),
            self.currentLinkage,
            None,
            ctx,
        )

        for i in range(len(ctx.params().param())):
            name, datatype = self.visit(ctx.params().param()[i])
            symbol.type.functionParameters.append((name, datatype))

        self.currentFunctionSymbols.append(symbol)
        self.visitChildren(ctx)
        self.currentFunctionSymbols.pop()

        if ctx.returntype():
            symbol.type.functionReturnType = self.visit(ctx.returntype())

        self.program.globalScope.defineSymbol(symbol, self.getLocation(ctx))
        self.program.externFunctionRefs.append(
            ExternFunctionRef(FunctionLinkage.External_C, self.getLocation(ctx), symbol)
        )

    def implFunc(
        self,
        ctx: (
            HazeParser.FuncContext
            | HazeParser.NamedfuncContext
            | HazeParser.StructFuncDeclContext
        ),
        name: str,
    ):
        scope = self.db.pushScope(
            Scope(self.getLocation(ctx), self.db.getCurrentScope())
        )

        if self.currentLinkage:
            raise ImpossibleSituation()

        symbol = FunctionSymbol(
            name,
            self.structStack[-1] if len(self.structStack) > 0 else None,
            Datatype.createFunctionType([], Datatype.createDeferredType()),
            FunctionLinkage.Haze,
            scope,
            ctx,
        )
        self.setNodeSymbol(ctx, symbol)

        for i in range(len(ctx.params().param())):
            nm, datatype = self.visit(ctx.params().param()[i])
            symbol.type.functionParameters.append((nm, datatype))
            if symbol.scope:
                symbol.scope.defineSymbol(
                    VariableSymbol(nm, symbol, datatype, VariableType.Parameter),
                    self.getLocation(ctx),
                )

        if ctx.returntype():
            symbol.type.functionReturnType = self.visit(ctx.returntype())
        else:
            symbol.type.functionReturnType = Datatype.createDeferredType()

        if symbol.type.functionReturnType.isDeferred():
            if ctx.funcbody().expr():
                expr = self.visit(ctx.funcbody().expr())
                symbol.type.functionReturnType = expr.type

        if symbol.parentSymbol and symbol.parentSymbol.type.isStruct():
            if name == "constructor":
                symbol.isConstructor = True
                if symbol.type.functionReturnType.isDeferred():
                    symbol.type.functionReturnType = symbol.parentSymbol.type
                else:
                    raise CompilerError(
                        f"Constructor of struct '{symbol.parentSymbol.name}' returns the struct itself implicitly",
                        self.getLocation(ctx),
                    )
            else:
                symbol.thisPointerType = Datatype.createPointerDatatype(
                    symbol.parentSymbol.type
                )
                if not symbol.scope:
                    raise ImpossibleSituation()
                symbol.scope.defineSymbol(
                    VariableSymbol(
                        "this", symbol, symbol.thisPointerType, VariableType.Parameter
                    ),
                    self.getLocation(ctx),
                )
        self.program.globalScope.defineSymbol(symbol, self.getLocation(ctx))

        self.db.popScope()
        return symbol

    def visitFunc(self, ctx):
        return self.implFunc(ctx, self.db.makeAnonymousFunctionName())

    def visitNamedfunc(self, ctx):
        return self.implFunc(ctx, ctx.ID().getText())

    def visitStructFieldDecl(self, ctx):
        name = ctx.ID().getText()
        datatype = self.visit(ctx.datatype())
        return VariableSymbol(
            name,
            self.structStack[-1],
            datatype,
            VariableType.MutableStructField,
        )

    def visitStructDecl(self, ctx):
        name = ctx.ID().getText()
        parentScope = self.db.getCurrentScope()
        scope = self.db.pushScope(Scope(self.getLocation(ctx), parentScope))

        genericsList = [(n.getText(), None) for n in ctx.datatype()]
        for generic in genericsList:
            scope.defineSymbol(
                GenericPlaceholderSymbol(generic[0]), self.getLocation(ctx)
            )

        symbol = DatatypeSymbol(
            name,
            self.structStack[-1] if len(self.structStack) > 0 else None,
            Datatype.createStructDatatype(name, genericsList),
        )
        parentScope.defineSymbol(symbol, self.getLocation(ctx))

        self.structStack.append(symbol)

        for content in ctx.structcontent():
            symbol.type.structMemberSymbols.insert(
                self.visit(content), self.getLocation(ctx)
            )

        self.structStack.pop()
        self.db.popScope()

    def visitStructFuncDecl(self, ctx):
        name = ctx.ID().getText()
        return self.implFunc(ctx, name)

File: src/SymbolCollector.ts
import {
  BooleanConstantContext,
  CdefinitiondeclContext,
  CommonDatatypeContext,
  FuncdeclContext,
  FunctionDatatypeContext,
  GenericsvalueContext,
  LiteralConstantContext,
  NamedfuncContext,
  NamespaceContext,
  ParamsContext,
  StringConstantContext,
  StructDeclContext,
  StructMemberContext,
  StructMethodContext,
  StructUnionFieldsContext,
  VariableDeclarationContext,
  VariableDefinitionContext,
  type ParamContext,
} from "./parser/grammar/autogen/HazeParser";
import type { Module } from "./Module";
import { HazeVisitor } from "./parser/grammar/autogen/HazeVisitor";
import { CompilerError, ImpossibleSituation, InternalError } from "./Errors";
import { ResolvedScope } from "./Scope";
import { Interval, type ParserRuleContext } from "antlr4ng";
import {
  collectFunction,
  collectVariableStatement,
  datatypeSymbolUsed,
  mangleSymbol,
  RESERVED_STRUCT_NAMES,
  visitBooleanConstantImpl,
  visitCommonDatatypeImpl,
  visitLiteralConstantImpl,
  visitParam,
  visitParams,
  visitStringConstantImpl,
  type ParamPack,
} from "./utils";
import type { Statement } from "./Statement";
import type { Parser } from "./parser";
import type { ParsedDatatype, ParsedSymbol } from "./ParsedTypes";
import { ELinkage, EVariableContext, EVariableMutability } from "./common";

export class SymbolCollector extends HazeVisitor<any> {
  private module: Module;
  private parentSymbolStack: ParsedSymbol.Symbol[];
  private parser?: Parser;

  constructor(module: Module) {
    super();
    this.module = module;
    this.parentSymbolStack = [];
  }

  collect = (ctx: ParserRuleContext, parser: Parser, filename: string) => {
    this.module.filename = filename;
    this.parser = parser;
    this.visit(ctx);
    this.parser = undefined;
    this.module.filename = undefined;
  };

  visitParam = (ctx: ParamContext): [string, ParsedDatatype.Datatype] => {
    return visitParam(this, ctx);
  };

  visitParams = (ctx: ParamsContext): ParamPack => {
    return visitParams(this, ctx);
  };

  visitCommonDatatype = (
    ctx: CommonDatatypeContext,
  ): ParsedSymbol.DatatypeSymbol => {
    return visitCommonDatatypeImpl(this, this.module, ctx);
  };

  visitCdefinitiondecl = (ctx: CdefinitiondeclContext): void => {
    const text = JSON.parse(ctx.STRING_LITERAL().getText());
    this.module.cDefinitionDecl.push(text);
  };

  visitFunctionDatatype = (
    ctx: FunctionDatatypeContext,
  ): ParsedSymbol.DatatypeSymbol => {
    const params = this.visitParams(ctx.functype().params());
    const type: ParsedDatatype.Function = {
      variant: "Function",
      functionParameters: params.params,
      functionReturnType: this.visit(ctx.functype().datatype()).type,
      vararg: params.vararg,
    };
    const symbol: Symbol = {
      name: "",
      scope: this.module.currentScope,
      type: type,
      variant: "Datatype",
      export: false,
      location: this.module.location(ctx),
    };
    datatypeSymbolUsed(symbol, this.module);
    return symbol;
  };

  visitFuncdecl = (ctx: FuncdeclContext): void => {
    // const extern = Boolean(ctx._extern);

    const lang = ctx.externlang()?.getText().slice(1, -1);
    let functype = ELinkage.Internal;
    if (lang === "C") {
      functype = ELinkage.External_C;
    } else if (functype) {
      throw new CompilerError(
        `Extern Language '${lang}' is not supported.`,
        this.module.location(ctx),
      );
    }

    const signature = ctx.ID().map((n) => n.getText());
    if (signature.length > 1 && functype === ELinkage.External_C) {
      throw new CompilerError(
        "Extern C functions cannot be namespaced",
        this.module.location(ctx),
      );
    }

    if (signature.length > 1) {
      throw new InternalError(
        "Namespacing for external function is not implemented yet",
      );
    }

    const symbol = collectFunction(
      this.module,
      this,
      ctx,
      signature[0],
      this.parentSymbolStack[this.parentSymbolStack.length - 1],
      functype,
    );
    if (symbol.isExported) {
      this.module.exportSymbols.set(mangleSymbol(this.module, symbol), symbol);
    }
  };

  visitNamedfunc = (ctx: NamedfuncContext): ParsedSymbol.FunctionSymbol => {
    return collectFunction(
      this.module,
      this,
      ctx,
      ctx.ID().getText(),
      this.parentSymbolStack[this.parentSymbolStack.length - 1],
    );
  };

  visitStructMember = (ctx: StructMemberContext) => {
    const datatypeSymbol: ParsedSymbol.DatatypeSymbol = this.visit(
      ctx.datatype(),
    );
    const symbol = this.module.parsedStore.createSymbol((id) => ({
      id,
      variant: "Variable",
      name: ctx.ID().getText(),
      type: datatypeSymbol.type,
      variableType: EVariableMutability.MutableStructField,
      variableContext: EVariableContext.MemberOfStruct,
      isExported: false,
      location: this.module.location(ctx),
      linkage: ELinkage.Internal,
    }));
    return symbol;
  };

  // visitStructUnionFields = (
  //   ctx: StructUnionFieldsContext,
  // ): StructMemberUnion => {
  //   return {
  //     variant: "StructMemberUnion",
  //     symbols: ctx.structcontent().map((n) => this.visit(n)),
  //   };
  // };

  visitStructDecl = (ctx: StructDeclContext): ParsedSymbol.DatatypeSymbol => {
    const name = ctx.ID(0)!.getText();
    const parentScope = this.module.currentScope;
    const scope = this.module.pushScope(
      this.module.parsedStore.createScope(
        this.module.location(ctx),
        parentScope.id,
      ),
    );

    if (RESERVED_STRUCT_NAMES.includes(name)) {
      throw new CompilerError(
        `'${name}' is a reserved name`,
        this.module.location(ctx),
      );
    }

    const genericsList: string[] = ctx
      .ID()
      .slice(1)
      .map((n) => n.getText());

    const lang = ctx.externlang()?.getText().slice(1, -1);
    let language = ELinkage.Internal;
    if (lang === "C") {
      language = ELinkage.External_C;
    } else if (lang) {
      throw new CompilerError(
        `Extern Language '${lang}' is not supported.`,
        this.module.location(ctx),
      );
    }

    let exports = false;
    if (ctx._export_) {
      exports = true;
    }

    const type = this.module.parsedStore.createDatatype((id) => ({
      id,
      variant: "Struct",
      name: name,
      language: language,
      generics: genericsList,
      members: [],
      methods: [],
      parentSymbol:
        this.parentSymbolStack[this.parentSymbolStack.length - 1].id,
    }));
    const symbol = this.module.parsedStore.createSymbol((id) => ({
      id,
      variant: "Datatype",
      name: name,
      type: type.id,
      definedInScope: scope.id,
      parentSymbol:
        this.parentSymbolStack[this.parentSymbolStack.length - 1].id,
      isExported: exports,
      location: this.module.location(ctx),
    }));

    if (genericsList.length > 0) {
      if (!this.parser?.parser || !ctx.stop) {
        throw new ImpossibleSituation();
      }
      let originalText = this.parser.parser.tokenStream.getTextFromInterval(
        new Interval(ctx.start?.tokenIndex || 0, ctx.stop.tokenIndex),
      );
      if (originalText.startsWith("export ")) {
        originalText = originalText.replace("export ", "");
      }
      // symbol.originalGenericSourcecode = originalText;
    }

    parentScope.defineSymbol(symbol);
    // for (const [name, tp] of type.generics) {
    //   const sym: DatatypeSymbol = {
    //     name: name,
    //     variant: "Datatype",
    //     scope: scope,
    //     type: { variant: "Generic", name: name },
    //     export: false,
    //     location: symbol.location,
    //   };
    //   scope.defineSymbol(sym);
    // }
    this.parentSymbolStack.push(symbol);

    ctx.structcontent().forEach((c) => {
      const content: ParsedSymbol.VariableSymbol | ParsedSymbol.FunctionSymbol =
        this.visit(c);
      if (content.variant === "Variable") {
        type.members.push(content);
      } else if (content.variant === "Function") {
        if (symbol.type.language) {
          throw new CompilerError(
            `A declared struct cannot have methods`,
            symbol.location,
          );
        }
        type.methods.push(content);
      } else if (content.variant === "StructMemberUnion") {
        type.members.push(content);
      } else {
        throw new ImpossibleSituation();
      }
    });

    if (symbol.export) {
      this.module.exportSymbols.set(mangleSymbol(symbol), symbol);
    }

    this.parentSymbolStack.pop();
    this.module.popScope();
    return symbol;
  };

  visitStructMethod = (
    ctx: StructMethodContext,
  ): ParsedSymbol.FunctionSymbol => {
    const name = ctx.ID().getText();
    const symbol = collectFunction(
      this,
      ctx,
      name,
      this.module,
      this.parentSymbolStack[this.parentSymbolStack.length - 1],
    );
    return symbol;
  };

  visitNamespace = (ctx: NamespaceContext): ParsedSymbol.DatatypeSymbol => {
    const names = ctx.ID().map((c) => c.getText());

    let symbol: Symbol | undefined =
      this.parentSymbolStack[this.parentSymbolStack.length - 1];
    do {
      let insertionSymbol: Symbol | undefined;
      let insertionScope: ResolvedScope;
      if (!symbol) {
        symbol = this.module.currentScope.tryLookupSymbolHere(names[0]);
        insertionScope = this.module.currentScope;
      } else {
        if (symbol.type.variant !== "Namespace") {
          throw new ImpossibleSituation();
        }
        insertionScope = symbol.type.symbolsScope;
        insertionSymbol = symbol;
        symbol = symbol.type.symbolsScope.tryLookupSymbolHere(names[0]);
      }
      if (!symbol) {
        const newScope = new ResolvedScope(
          this.module.location(ctx),
          this.module.currentScope,
        );
        if (insertionSymbol && insertionSymbol.variant !== "Datatype") {
          throw new ImpossibleSituation();
        }
        symbol = {
          name: names[0],
          variant: "Datatype",
          scope: newScope,
          type: {
            variant: "Namespace",
            name: names[0],
            symbolsScope: newScope,
            parentSymbol: insertionSymbol,
          },
          parentSymbol: insertionSymbol,
          export: false,
          location: this.module.location(ctx),
        };
        insertionScope.defineSymbol(symbol);
      }
      names.splice(0, 1);
    } while (names.length > 0);

    if (symbol.variant !== "Datatype" || symbol.type.variant !== "Namespace") {
      throw new InternalError("Namespace is not a datatype");
    }

    this.module.pushScope(symbol.scope);
    this.parentSymbolStack.push(symbol as ParsedSymbol.DatatypeSymbol);

    ctx.namespacecontent().children?.forEach((n) => {
      this.visit(n);
    });

    this.parentSymbolStack.pop();
    this.module.popScope();

    if (ctx._export_) {
      this.module.exportSymbols.set(mangleSymbol(this.module, symbol), symbol);
    }

    return symbol;
  };

  visitGenericsvalue = (
    ctx: GenericsvalueContext,
  ): ConstantSymbol | DatatypeSymbol => {
    if (ctx.constant()) {
      return this.visit(ctx.constant()!) as ConstantSymbol;
    } else {
      return this.visit(ctx.datatype()!) as DatatypeSymbol;
    }
  };

  visitStringConstant = (ctx: StringConstantContext): StringConstantSymbol => {
    return visitStringConstantImpl(this, ctx, this.module);
  };

  visitLiteralConstant = (
    ctx: LiteralConstantContext,
  ): ParsedSymbol.ConstantSymbol => {
    return visitLiteralConstantImpl(this, ctx, this.module);
  };

  visitBooleanConstant = (
    ctx: BooleanConstantContext,
  ): ParsedSymbol.BooleanConstantSymbol => {
    return visitBooleanConstantImpl(this, ctx, this.module);
  };

  visitVariableDefinition = (ctx: VariableDefinitionContext): Statement => {
    return collectVariableStatement(
      this,
      ctx,
      this.module,
      VariableScope.Global,
      this.parentSymbolStack[this.parentSymbolStack.length - 1],
    );
  };

  visitVariableDeclaration = (ctx: VariableDeclarationContext): Statement => {
    return collectVariableStatement(
      this,
      ctx,
      this.module,
      VariableScope.Global,
      this.parentSymbolStack[this.parentSymbolStack.length - 1],
    );
  };
}
